src/graph.py

```diff
@@ -0,0 +1,174 @@
+            for i in values:
+                to_return.append([keys, i[0], i[1]])
+        return to_return
+
+    def del_node(self, val):
+        """Deletes a node from the graph, and from all edge pointers."""
+        try:
+            del self.graph_dict[val]
+            for keys, values in self.graph_dict.items():
+                for i in values:
+                    if i[0] == val:
+                        values.remove(i)
+        except KeyError:
+            raise KeyError("No such node exists.")
+    def del_edge(self, val1, val2):
+        """Deletes an edge from graph."""
+        try:
+            for node in self.graph_dict[val1]:
+                if node[0] == val2:
+                    self.graph_dict[val1].remove(node)
+        except ValueError:
+            raise ValueError("No such edge exists.")
+        except KeyError:
+            raise KeyError("No such node exists.")
+    def has_node(self, val):
+        """Checks if graph has a given node in it."""
+        try:
+            self.graph_dict[val]
+            return True
+        except KeyError:
+            return False
+    def neighbors(self, val):
+        """Return all nodes connected to given node."""
+        try:
+            to_return = []
+            for value in self.graph_dict[val]:
+                to_return.append(value)
+            return to_return
+        except KeyError:
+            raise KeyError("No such node exists.")
+    def adjacent(self, val1, val2):
+        """Return True if edge exists, else return false."""
+        try:
+            self.graph_dict[val2]
+            for node in self.graph_dict[val1]:
+                if node[0] == val2:
+                    return True
+            else: 
+                return False
+        except KeyError:
+            raise KeyError("Value given not in graph.")
+    def depth(self, val):
+        from stack import Stack
+        seen = []
+        next_up = Stack()
+        try:
+            while True:
+                if val not in seen:
+                    seen.append(val)
+                    for i in self.graph_dict[val][::-1]:
+                        next_up.push(i)
+                if len(next_up) == 0:
+                    break
+                val = next_up.pop().value[0]
+            return seen
+        except KeyError:
+            raise KeyError('Given value does not exist.')
+    def breadth(self, val):
+        from que_ import Queue
+        seen = []
+        next_up = Queue()
+        try:
+            while True:
+                if val not in seen:
+                    seen.append(val)
+                    for i in self.graph_dict[val]:
+                        next_up.enqueue(i)
+                if next_up.size() == 0:
+                    break
+                val = next_up.dequeue().value[0]
+            return seen
+        except KeyError:
+            raise KeyError('Given value does not exist.')
+    def path(self, val1):
```

Comment:
Methods that shouldn't be public and are helper functions should start with an underscore